File: src/main/java/se/andolf/CPU.java
package se.andolf;

import se.andolf.utils.Utils;

import java.util.Random;
import java.util.Stack;

public class CPU {

    private final Memory memory;
    private final Register register;
    private final Random random;
    private final Keyboard keyboard;
    private final Screen screen;
    private final Stack<Integer> stack;

    private boolean drawFlag;

    private int I;
    private int pc;
    private int delayTimer, soundTimer;

    public CPU(Memory memory, Register register, Random random, Keyboard keyboard, Screen screen) {
        this.memory = memory;
        this.register = register;
        this.random = random;
        this.keyboard = keyboard;
        this.screen = screen;

        stack = new Stack<>();

        pc = 0x200;
        I = 0;
    }

    public void emulateCycle() {

        final int opcode = memory.getOpcode(pc);
        System.out.println(Utils.getHex(opcode));


        int x = (opcode & 0x0F00) >>> 8;
        int y = (opcode & 0x00F0) >>> 4;

        switch(opcode) {
            case 0x00E0:
                screen.clear();
                drawFlag = true;
                pc += 2;
                return;

            case 0x00EE:
                pc = stack.pop();
                drawFlag = true;
                pc += 2;
                break;
        }

        switch (opcode & 0xF000) {

            case 0x1000:
                pc = opcode & 0x0FFF;
                break;

            case 0x2000:
                stack.push(pc);
                pc = opcode & 0x0FFF;
                break;

            case 0x3000:
                if(register.get(x) == (opcode & 0x00FF)) {
                    pc += 4;
                } else {
                    pc += 2;
                }
                break;

            case 0x4000:
                if(register.get((opcode & 0x0F00) >>> 8) != (opcode & 0x00FF)) {
                    pc += 4;
                } else {
                    pc += 2;
                }
                break;

            case 0x5000:
                if(register.get((opcode & 0x0F00) >>> 8) == register.get((opcode & 0x00F0) >>> 4)) {
                    pc += 4;
                } else {
                    pc += 2;
                }
                break;

            case 0x6000:
                register.set((opcode & 0x0F00) >>> 8, (opcode & 0x00FF));
                pc += 2;
                break;

            case 0x7000:
                register.apply(x, xv -> {
                    int result = xv + (opcode & 0x00FF);
                    if(result >= 256) {
                        return (result - 256);
                    } else {
                        return result;
                    }
                });
                pc += 2;
                break;
        }


        switch (opcode & 0xF00F) {

            case 0x8000:
                register.set(x, y);
                pc += 2;
                break;

            case 0x8001:
                register.apply(x, vx -> vx | register.get(y));
                pc += 2;
                break;

            case 0x8002:
                register.apply(x, vx -> vx & register.get(y));
                pc += 2;
                break;

            case 0x8003:
                register.apply(x, vx -> vx ^ register.get(y));
                pc += 2;
                break;

            case 0x8004:
                register.apply(x, operand -> {
                    int sum = operand + register.get(y);
                    register.set(0xF, sum > 0xFF ? 1 : 0);
                    return (sum & 0xFF);
                });
                pc += 2;
                break;

            case 0x8005:
                register.apply(x, vx -> {
                    register.set(0xF, register.get(y) > vx ? 0 : 1);
                    return (vx - register.get(y)) & 0xFF;
                });
                pc += 2;
                break;

            case 0x8006:
                register.set(0xF, (register.get(x) & 0x1));
                register.apply(x, vx -> vx >>> 1);
                pc += 2;
                break;

            case 0x8007:
                register.set(0xF, register.get(y) < register.get(x) ? 0 : 1);
                register.apply(x, vx -> (register.get(y) - vx) & 0xFF);
                pc += 2;
                break;

            case 0x800E:
                register.set(0xF, (register.get(x) & 0x80) > 0 ? 1 : 0);
                register.apply(x, vx -> (vx << 1) & 0xFF);
                pc += 2;
                break;

            case 0x9000:
                if(register.get(x) != register.get(y))
                    pc += 4;
                else
                    pc += 2;
                break;

        }

        switch (opcode & 0xF000) {
            case 0xA000:
                I = opcode & 0x0FFF;
                pc += 2;
                break;

            case 0xB000:
                pc = (opcode & 0x0FFF) + register.get(0);
                break;

            case 0xC000:
                register.set(x, random.nextInt(256) & (opcode & 0x00FF));
                pc += 2;
                break;

            case 0xD000:

                final int height = opcode & 0x000F;

                register.set(0xF, 0);

                for (int yLine = 0; yLine < height; yLine++) {

                    int pixelValue = memory.getByte(I + yLine);

                    for (int xLine = 0; xLine < 8; xLine++) {

                        if (Utils.getBitValue(pixelValue, xLine) != 0) {

                            int xCoord = (register.get(x) + xLine);
                            int yCoord = (register.get(y) + yLine) ;

                            if(xCoord >= screen.getWidth() )
                                xCoord %= screen.getWidth();

                            if(yCoord >= screen.getHeight())
                                yCoord %= screen.getHeight();

                            if (screen.getPixel(xCoord, yCoord) == 1)
                                register.set(0xF, 1);

                            screen.setPixel(xCoord, yCoord);
                        }
                    }
                }
                drawFlag = true;
                pc += 2;
                break;
        }

        switch (opcode & 0xF0FF) {

            case 0xE09E:
                if(keyboard.isPressed(register.get(x))) {
                    pc += 4;
                } else {
                    pc += 2;
                }
                break;

            case 0xE0A1:
                if(!keyboard.isPressed(register.get(x))) {
                    pc += 4;
                } else {
                    pc += 2;
                }
                break;

            case 0xF00A:
                for (int i = 0; i < keyboard.getKeys().length; i++) {
                    if(keyboard.isPressed(i)){
                        register.set(x, i);
                        pc += 2;
                    }
                }
                break;

            case 0xF007:
                register.set(x, this.delayTimer);
                pc += 2;
                break;

            case 0xF015:
                this.delayTimer = register.get(x);
                pc += 2;
                break;

            case 0xF018:
                this.soundTimer = register.get(x);
                pc += 2;
                break;

            case 0xF01E:
                this.I += register.get(x);
                pc += 2;
                break;

            case 0xF029:
                this.I = register.get(x) * 5;
                pc += 2;
                break;

            case 0xF033:

                memory.setByte(I, register.get(x) / 100);
                memory.setByte(I + 1, (register.get(x) % 100) / 10);
                memory.setByte(I + 2, (register.get(x) % 100) % 10);

                pc += 2;
                return;

            case 0xF055:
                for (int j = 0; j <= x; j++) {
                    memory.setByte(I + j, register.get(j));
                }

                pc += 2;
                return;

            case 0xF065:
                for (int j = 0; j <= x; j++) {
                    register.set(j, memory.getByte(I + j) & 0xFF);
                }
                pc += 2;
                break;
        }
    }

    public int getDelayTimer() {
        return delayTimer;
    }

    public int getSoundTimer() {
        return soundTimer;
    }

    public void setDelayTimer(int value) {
        delayTimer = value;
    }

    public boolean getDrawFlag() {
        return drawFlag;
    }

    public void setDrawFlag(boolean value) {
        drawFlag = value;
    }

    public int getPC() {
        return pc;
    }

    public int getI() {
        return I;
    }
}

File: src/main/dotnet/Emulator/CPU.cs
namespace Chip8;

public class CPU 
{
    private readonly Memory memory;
    private readonly Register register;
    private readonly Random random;
    private readonly IKeyboard keyboard;
    private readonly IScreen screen;
    private Stack<int> stack;

    private bool drawFlag;

    private int i;
    private int pc;
    private byte delayTimer, soundTimer;
    public int InstructionsCounter;

    public CPU(Memory memory, Register register, Random random, IKeyboard keyboard, IScreen screen) {
        this.memory = memory;
        this.register = register;
        this.random = random;
        this.keyboard = keyboard;
        this.screen = screen;

        stack = new Stack<int>();

        pc = 0x200;
        i = 0;
    }


    public void EmulateCycle() 
    {
        InstructionsCounter++;
        int opcode = memory.GetOpcode(pc);
        //Console.WriteLine(opcode.ToString("X"));

        int x = (opcode & 0x0F00) >> 8;
        int y = (opcode & 0x00F0) >> 4;
        byte VX = register.Get(x);
        byte VY = register.Get(y);
        byte N = (byte)(opcode & 0x000F);
        byte NN = (byte)(opcode & 0x00FF);
        int NNN = (opcode & 0x0FFF);

        pc += 2;

        switch(opcode) {
 
            case 0x00E0: // Clears the screen
                screen.Clear();
                drawFlag = true;
                return;
 
            case 0x00EE: // Returns from a subroutine.
                pc = stack.Pop();
                return;
 
            default:
                break;
        }

        switch(opcode & 0xF000) {
 
            case 0x1000: // Jumps to address NNN.
                pc = NNN;
                break;
 
            case 0x2000: // Calls subroutine at NNN.
                stack.Push(pc);
                pc = NNN;
                break; 

            case 0x3000: // Skips the next instruction if VX equals NN. (Usually the next instruction is a jump to skip a code block)
                if(VX == NN) {
                    pc += 2;
                }
                break;

            case 0x4000: // Skips the next instruction if VX doesn't equal NN. (Usually the next instruction is a jump to skip a code block)
                if(VX != NN) {
                    pc += 2;
                }
                break;

            case 0x5000: // Skips the next instruction if VX equals VY. (Usually the next instruction is a jump to skip a code block)
                if(VX == VY) {
                    pc += 2;
                }
                break;

            case 0x6000: // Sets VX to NN.
                register.Set(x, NN);
                break;

            case 0x7000: // Adds NN to VX. (Carry flag is not changed)
                register.Apply(x, vx => (byte)((vx + NN) & 0xFF));
                break;

            case 0x8000:{
                switch(N) {
                    
                    case 0x0000: // Sets VX to the value of VY.
                        register.Set(x, VY);
                        break;

                    case 0x0001: //	Sets VX to VX or VY. (Bitwise OR operation)
                        register.Apply(x, vx => (byte)(vx | VY));
                        break;

                    case 0x0002: // Sets VX to VX and VY. (Bitwise AND operation)
                        register.Apply(x, vx => (byte)(vx & VY));
                        break;

                    case 0x0003: // Sets VX to VX xor VY.
                        register.Apply(x, vx => (byte)(vx ^ VY));
                        break;

                    case 0x0004: // Adds VY to VX. VF is set to 1 when there's a carry, and to 0 when there isn't.
                        register.Apply(x, vx => {
                            var sum = vx + VY;
                            register.Set(0xF, (byte)(sum > 0xFF ? 1 : 0));
                            return (byte)(sum & 0xFF);
                        });
                        break;

                    case 0x0005: // VY is subtracted from VX. VF is set to 0 when there's a borrow, and 1 when there isn't.
                        register.Set(0xF, (byte)(VY > VX ? 0 : 1));
                        register.Apply(x, vx => (byte)((vx - VY) & 0xFF));
                        break;

                    case 0x0006: // Stores the least significant bit of VX in VF and then shifts VX to the right by 1.[2]
                        register.Set(0xF, (byte)(VX & 0x1));
                        register.Apply(x, vx => (byte)(vx >> 1));
                        break;

                    case 0x0007: // Sets VX to VY minus VX. VF is set to 0 when there's a borrow, and 1 when there isn't.
                        register.Set(0xF, (byte)(VX > VY ? 0 : 1));
                        register.Apply(x, vx => (byte)((VY - vx) & 0xFF));
                        break;

                    case 0x000E: // Stores the most significant bit of VX in VF and then shifts VX to the left by 1.[3]
                        register.Set(0xF, (byte)((VX & 0x80) > 0 ? 1 : 0));
                        register.Apply(x, vx => (byte)((vx << 1) & 0xFF));
                        break;
                    default:
                        break;
                    }
                }
                break;

            case 0x9000: // Skips the next instruction if VX doesn't equal VY. (Usually the next instruction is a jump to skip a code block)
                pc += VX != VY ? 2 : 0;
                break;

            case 0xA000: // Sets I to the address NNN.
                i = NNN;
                break;

            case 0xB000: // Jumps to the address NNN plus V0..
                pc = (register.Get(0) + NNN) & 0x0FFF;
                break;

            case 0xC000: // Sets VX to the result of a bitwise and operation on a random number (Typically: 0 to 255) and NN.
                register.Set(x, (byte)((random.Next() % 256) & NN));
                break;
            
            case 0xD000: // Draws a sprite at coordinate (VX, VY) that has a width of 8 pixels and a height of N pixels. Each row of 8 pixels is read as bit-coded starting from memory location I; I value doesn’t change after the execution of this instruction. As described above, VF is set to 1 if any screen pixels are flipped from set to unset when the sprite is drawn, and to 0 if that doesn’t happen
                int height = N;

                register.Set(0xF, 0);

                for (int yLine = 0; yLine < height; yLine++) {
                    int pixelValue = memory.GetByte(I + yLine);

                    for (int xLine = 0; xLine < 8; xLine++) {

                        if (Utils.GetBitValue(pixelValue, xLine) != 0) {

                            int xCoord = (VX + xLine);
                            int yCoord = (VY + yLine) ;

                            if(xCoord >= screen.Width )
                                xCoord %= screen.Width;

                            if(yCoord >= screen.Height)
                                yCoord %= screen.Height;

                         if (screen.GetPixel(xCoord, yCoord) == 1)
                                register.Set(0xF, 1);

                            screen.SetPixel(xCoord, yCoord);
                        }
                    }
                }
                drawFlag = true;
                break;
            
            case 0xE000: {
                switch (NN) {
                    
                    case 0x9E: // Skips the next instruction if the key stored in VX is pressed. (Usually the next instruction is a jump to skip a code block)
                        pc += keyboard.IsPressed(VX) ? 2 : 0;
                        break;

                    case 0xA1: // Skips the next instruction if the key stored in VX isn't pressed. (Usually the next instruction is a jump to skip a code block)
                        pc += keyboard.IsPressed(VX) ? 0 : 2;
                        break;

                    default:
                        break;
                }
                break;
            }

            case 0xF000: {
                switch (NN) {

                    case 0x07: // Sets VX to the value of the delay timer.
                        register.Set(x, (byte)(delayTimer & 0xFF));
                        break;

                    case 0x0A: // A key press is awaited, and then stored in VX. (Blocking Operation. All instruction halted until next key event)
                        for (int i = 0; i < keyboard.GetKeys().Length; i++) {
                            if(keyboard.IsPressed(i)){
                                register.Set(x, (byte)i);
                                break;
                            }
                        }
                        pc -= 2;
                        break;

                    case 0x15: // Sets the delay timer to VX.
                        delayTimer = (byte)(VX & 0xFF);
                        return;

                    case 0x18: // Sets the sound timer to VX.
                        soundTimer = (byte)(VX & 0xFF);
                        return;

                    case 0x1E: // Adds VX to I.
                        i += (VX & 0xFF);
                        break;

                    case 0x29: // Sets I to the location of the sprite for the character in VX. Characters 0-F (in hexadecimal) are represented by a 4x5 font.
                        i = 0x50 + VX * 5;
                        break;

                    case 0x33: // VX, with the most significant of three digits at the address in I, the middle digit at I plus 1, and the least significant digit at I plus 2. (In other words, take the decimal representation of VX, place the hundreds digit in memory at location in I, the tens digit at location I+1, and the ones digit at location I+2.)
                        memory.SetByte(i, (Byte)(VX / 100));
                        memory.SetByte(i + 1, (Byte)((VX % 100) / 10));
                        memory.SetByte(i + 2, (Byte)(VX % 10));
                        break;

                    case 0x55: // Stores V0 to VX (including VX) in memory starting at address I. The offset from I is increased by 1 for each value written, but I itself is left unmodified.
                        for(int p = 0; p <= x; p++)
                            memory.SetByte(i + p, (Byte)register.Get(p));
                        break;

                    case 0x65: // Fills V0 to VX (including VX) with values from memory starting at address I. The offset from I is increased by 1 for each value written, but I itself is left unmodified.
                        for(int p = 0; p <= x; p++)
                            register.Set(p, (byte)(memory.GetByte(i + p) & 0xFF));
                        break;

                    default:
                        break;
                    }
                }
                break;

            default:
                break;
        }

        if(this.delayTimer > 0) {
            this.delayTimer -= 1;
        }

        if(this.soundTimer > 0) {
            this.soundTimer -= 1;
        }
    }

    public byte DelayTimer { get { return delayTimer; } set { delayTimer = value; } }
    public bool DrawFlag { get { return drawFlag; } set { drawFlag = value; } }

    public byte SoundTimer => soundTimer;

    public int PC { get { return pc; } }

    public int I { get { return i; } }
}
